datafusion/common/src/scalar.rs

```diff
@@ -2073,14 +2082,14 @@ impl ScalarValue {
         value: &Option<i128>,
         precision: u8,
         scale: u8,
-    ) -> bool {
-        let array = array.as_any().downcast_ref::<Decimal128Array>().unwrap();
+    ) -> Result<bool> {
+        let array = as_decimal128_array(array)?;
         if array.precision() != precision || array.scale() != scale {
-            return false;
+            return Ok(false);
         }
         match value {
```

Comment:
Using `match` expressions with binary types like `Result` or `Option` is typically less idiomatic than using the `if let` expression. With this in mind, I suggest:
```rust
let is_null = array.is_null(index);
if let Some(v) = value {
  Ok(!is_null && array.value(index) == *v)
} else {
  Ok(is_null)
}
```